packages/multi-test/src/app.rs

```diff
@@ -19,153 +19,345 @@ use crate::wasm::{ContractData, Wasm, WasmKeeper};
 use crate::BankKeeper;
 
 use anyhow::Result as AnyResult;
-use derivative::Derivative;
 
 pub fn next_block(block: &mut BlockInfo) {
     block.time = block.time.plus_seconds(5);
     block.height += 1;
 }
 
+/// Type alias for default build `App` to make its storing simpler in typical scenario
+pub type BasicApp<ExecC = Empty, QueryC = Empty> = App<
+    BankKeeper,
+    MockApi,
+    MockStorage,
+    PanickingCustomHandler<ExecC, QueryC>,
+    WasmKeeper<ExecC, QueryC>,
+>;
+
 /// Router is a persisted state. You can query this.
 /// Execution generally happens on the RouterCache, which then can be atomically committed or rolled back.
 /// We offer .execute() as a wrapper around cache, execute, commit/rollback process.
-///
-/// ExecC is the custom message returned init, handle, sudo (Response<C>).
-/// All contracts must return Response<C> or Response<Empty>.
-///
-/// Also `ExecC` is the custom message which is handled by custom message handler.
-///
-/// `QueryC` is custom query message handled by custom message handler.
-pub struct App<ExecC = Empty, QueryC = Empty>
+pub struct App<
+    Bank = BankKeeper,
+    Api = MockApi,
+    Storage = MockStorage,
+    Custom = PanickingCustomHandler<Empty, Empty>,
+    Wasm = WasmKeeper<Empty, Empty>,
+> {
+    router: Router<Bank, Custom, Wasm>,
+    api: Api,
+    storage: Storage,
+    block: BlockInfo,
+}
+
+impl Default for BasicApp {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl BasicApp {
+    /// Creates new default `App` implementation working with Empty custom messages.
+    pub fn new() -> Self {
+        AppBuilder::new().build()
+    }
+}
+
+/// Creates new default `App` implementation working with customized exec and query messages.
+/// Outside of `App` implementation to make type elision better.
+pub fn custom_app<ExecC, QueryC>() -> BasicApp<ExecC, QueryC>
 where
-    ExecC: Clone + fmt::Debug + PartialEq + JsonSchema + 'static,
+    ExecC: Debug + Clone + PartialEq + JsonSchema + DeserializeOwned + 'static,
+    QueryC: Debug + CustomQuery + DeserializeOwned + 'static,
 {
-    router: Router<ExecC, QueryC>,
-    api: Box<dyn Api>,
-    storage: Box<dyn Storage>,
-    block: BlockInfo,
+    AppBuilder::new_custom().build()
 }
 
-impl<ExecC, QueryC> Querier for App<ExecC, QueryC>
+impl<BankT, ApiT, StorageT, CustomT, WasmT> Querier for App<BankT, ApiT, StorageT, CustomT, WasmT>
 where
-    ExecC: Clone + fmt::Debug + PartialEq + JsonSchema + 'static,
-    QueryC: CustomQuery + DeserializeOwned,
+    CustomT::ExecC: Clone + fmt::Debug + PartialEq + JsonSchema + DeserializeOwned + 'static,
```

Comment:
Eyes starting to hurt....

You are a true soldier to make it to the end 🎖️

Reply:
This actually can be simplified - even if simply introducing marker trait:
```rust
trait CustomExecMsg: Clone + fmt::Debug + PartialEq + JsonSchema + DeserializeOwned + 'static {}

impl<C> CustomExecMsg for C where C: Clone + fmt::Debug + PartialEq + JsonSchema + DeserializeOwned + 'static {}
```

And then it can be used in all those long bounds. Also it is a trick for reducing all those app-traits to one, so instead of:
```rust
where
  BankT: Bank,
  CustomT: CustomHandler<ExecC = ExecC, QueryC = QueryC>,
  WasmT: Wasm<ExecC, QueryC>,
```

you would have:

```rust
where
  Self: App<BankT, CustomT, WasmT>
```

By extracting everything from `impl App` to separate trait `App` (and renaming `App` to some `AppImpl`) - bounds would be only in place of this trait impl. Same could be done for router (which is often bound in `Wasm`).

However I didn't want to include it in this review - it is big enough, and I didn't want to risk ruining something with those things.